### PR TITLE
Added ash_window::get_present_support

### PR DESCRIPTION
**Pull request by Rob2309**
This PR introduces `ash_window::get_present_support`, which is a higher-level wrapper around the various platform-specific functions for querying whether a physical device queue family supports presenting to surfaces (see [the vulkan spec](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap34.html#platformQuerySupport_android)).

This allows choosing a physical device before creating a surface.

### CODE REVIEW

File: ash-window/src/lib.rs [hunk 1]
```diff
@@ -171,3 +171,69 @@ pub fn enumerate_required_extensions(
...
+            ))
+        },
+        RawDisplayHandle::Xcb(h) => unsafe {
+            // https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap34.html#platformQuerySupport_xcb
+            let ext = khr::XcbSurface::new(entry, instance);
+            Ok(ext.get_physical_device_xcb_presentation_support(
+                physical_device,
+                queue_family_index,
+                &mut *h.connection,
+                h.screen as _,
```
Comment (Ralith) [inline ash-window/src/lib.rs:223]: This is supposed to be a visual, not a screen.

File: ash-window/src/lib.rs [hunk 2]
```diff
@@ -171,3 +171,69 @@ pub fn enumerate_required_extensions(
...
+            ))
+        },
+        RawDisplayHandle::Xlib(h) => unsafe {
+            // https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap34.html#platformQuerySupport_xlib
+            let ext = khr::XlibSurface::new(entry, instance);
+            Ok(ext.get_physical_device_xlib_presentation_support(
+                physical_device,
+                queue_family_index,
+                h.display,
+                h.screen as _,
```
Comment (Ralith) [inline ash-window/src/lib.rs:233]: As above.

**Review by Ralith — CHANGES_REQUESTED** [review]
Thanks, this looks really useful!

File: ash-window/src/lib.rs (hunk 2 above)
Comment (Rob2309) [reply ash-window/src/lib.rs:233]: You're right, I think a visualID can be somehow derived from the screen, but I think this would require linking to some X library. As an alternative, the get_present_support function could require a window handle, but this would defeat the main use case.

Comment (Ralith) [reply ash-window/src/lib.rs:233]: Yeah. We could pull in `x11-dl` and use it to make the necessary call (`XDefaultVisualOfScreen` I guess?). Bonus points for adding a cargo feature to use `x11` instead, as dynamic linking is a bit tidier than runtime loading on Linux, but it can be nice to avoid a buildtime dep on a foreign lib.

Alternatively, maybe a more convenient approach would be adding a visual_id member to  to `raw_window_handle::XlibDisplayHandle` upstream (and making sure winit populates it). This is the same motivation behind its inclusion in the window handle struct, though I'm not enough of an X11 expert to judge how weird that might be.

File: ash-window/src/lib.rs [hunk 3]
```diff
@@ -171,3 +171,69 @@ pub fn enumerate_required_extensions(
 
     Ok(extensions)
 }
+
+/// Query whether a `queue_family` of the given `physical_device` supports presenting to any surface that might be created.
+/// This function can be used to find a suitable [`vk::PhysicalDevice`] and queue family
+/// for rendering before a single surface is created.
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:177]: Nit: Can you reflow these sentences to both be approx the same length (line length is 100 characters): https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#max_width

File: ash-window/src/lib.rs [hunk 4]
```diff
@@ -171,3 +171,69 @@ pub fn enumerate_required_extensions(
...
+///
+/// This function can be a more useful alternative for
+/// [`VkGetPhysicalDeviceSurfaceSupportKHR`](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap34.html#vkGetPhysicalDeviceSurfaceSupportKHR),
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:180]: 1. This is a function, and starts with a _lowercase_ `v`;
2. It is preferred to have (long) links at the bottom, so that the source text isn't too crowded and oddly flowed (``[`vkGetPhysicalDeviceSurfaceSupportKHR`]``) automatically crates a label that you can assign a link to via ``/// [`vkGetPhysicalDeviceSurfaceSupportKHR`]: https://...``
   (More important below);
3. Since Ash is a dependency (and in scope via `ash::extensions::khr`), prefer linking to the Rust implementation of this function via ``[`khr::Surface::get_physical_device_surface_support()`]``.

(And separately link the docs page _if_ there's more useful information to describe explicitly)

File: ash-window/src/lib.rs [hunk 5]
```diff
@@ -171,3 +171,69 @@ pub fn enumerate_required_extensions(
...
+/// which requires having an actual surface available before choosing a physical device.
+///
+/// For more information see [the vulkan spec on WSI integration](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap34.html#_querying_for_wsi_support).
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:183]:
```suggestion
/// For more information see [the Vulkan spec on WSI integration][_querying_for_wsi_support].
///
/// [_querying_for_wsi_support]: https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap34.html#_querying_for_wsi_support
```

File: ash-window/src/lib.rs [hunk 6]
```diff
@@ -171,3 +171,69 @@ pub fn enumerate_required_extensions(
...
+            // On Android, iOS and macOS, every queue family supports presenting to any surface
+            Ok(true)
+        }
+        RawDisplayHandle::Wayland(h) => unsafe {
+            // https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap34.html#platformQuerySupport_walyand
+            let ext = khr::WaylandSurface::new(entry, instance);
+            Ok(ext.get_physical_device_wayland_presentation_support(
+                physical_device,
+                queue_family_index,
+                &mut *h.display.cast(),
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:205]: This doesn't look correct, this function takes `&mut vk::wl_display` (which is a type alias to `c_void`) but I think it should take `*mut vk::wl_display`; should we make a breaking change to improve that?

File: ash-window/src/lib.rs [hunk 7]
```diff
@@ -171,3 +171,69 @@ pub fn enumerate_required_extensions(
...
+                queue_family_index,
+            ))
+        },
+        RawDisplayHandle::Xcb(h) => unsafe {
+            // https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap34.html#platformQuerySupport_xcb
+            let ext = khr::XcbSurface::new(entry, instance);
+            Ok(ext.get_physical_device_xcb_presentation_support(
+                physical_device,
+                queue_family_index,
+                &mut *h.connection,
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:222]: Same here, I think we should take `*mut vk::xcb_connection_t`.

File: ash-window/src/lib.rs [hunk 8]
```diff
@@ -171,3 +171,69 @@ pub fn enumerate_required_extensions(
...
+                h.screen as _,
+            ))
+        },
+        RawDisplayHandle::Xlib(h) => unsafe {
+            // https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap34.html#platformQuerySupport_xlib
+            let ext = khr::XlibSurface::new(entry, instance);
+            Ok(ext.get_physical_device_xlib_presentation_support(
+                physical_device,
+                queue_family_index,
+                h.display,
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:232]: I.e. this is a `*mut vk::Display`, not a `&mut `vk::Display` :grimacing:

File: ash-window/src/lib.rs [hunk 9]
```diff
@@ -171,3 +171,69 @@ pub fn enumerate_required_extensions(
...
+                h.screen as _,
+            ))
+        },
+        // All other platforms mentioned in the vulkan spec are not supported by ash.
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:236]:
```suggestion
        // All other platforms mentioned in the Vulkan spec are not supported by ash-window.
```

Because `ash` maps the entire Vulkan API (== supports everything), even though we might not have written a slightly-higher-level `Fn` wrapper for it yet the lower-level API is still available.

Though I would reword that to `... don't currently have an implementation in ash-window`, as we can always expand.

**Review by MarijnS95 — CHANGES_REQUESTED** [review]

**Comment by MarijnS95** [comment]
While reviewing and refactoring the PR message, should we also mention (in the PR/commit message _and the doc-comment_) that it checks for support on the given `Display` connection? I'm not sure how much of a difference it makes beyond allowing the Vulkan driver to connect to the WSI though (perhaps for Xorg `Display` handles that could be networked?).

**Comment by Rob2309** [comment]
I think I addressed everything above in this new commit.
I added x11 and x11-dl as optional dependencies, x11-dl being enabled by default. There does not seem to be a xcb-dl lib so I added xcb as a mandatory dependency for now.
The xlib code path is working on my wsl2 installation, as for xcb I am not sure, winit seems to not actually support xcb as far as I can tell.
I also changed the parameters of `get_physical_device_xcb_presentation_support` and `get_physical_device_wayland_presentation_support` to be pointers instead of references.

**Comment by Rob2309** [comment]
I just noticed that `--no-default-features` seems to be ignored for running the winit example, I am not sure if I made a mistake or this is a cargo quirk.

File: ash/src/extensions/khr/wayland_surface.rs [hunk 10]
```diff
@@ -43,7 +43,7 @@ impl WaylandSurface {
         &self,
         physical_device: vk::PhysicalDevice,
         queue_family_index: u32,
-        wl_display: &mut vk::wl_display,
+        wl_display: *mut vk::wl_display,
```
Comment (MarijnS95) [inline ash/src/extensions/khr/wayland_surface.rs:46]: We should write these down as breaking changes in the `CHANGELOG` (maybe via a separate PR so that it shows up properly in the commit history), though if someone passes exactly a `&mut vk::wl_display` it'll automatically coerce to the raw pointer.

File: ash-window/src/lib.rs [hunk 11]
```diff
@@ -171,3 +171,102 @@ pub fn enumerate_required_extensions(
 
     Ok(extensions)
 }
+
+/// Query whether a `queue_family` of the given `physical_device` supports presenting to any
+/// surface that might be created. This function can be used to find a suitable
+/// [`vk::PhysicalDevice`] and queue family for rendering before a single surface is created.
+///
+/// This function can be a more useful alternative for [`vkGetPhysicalDeviceSurfaceSupportKHR`],
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:179]: I meant to just have:
```suggestion
/// This function can be a more useful alternative for [`khr::Surface::get_physical_device_surface_support()`],
```

We don't generally use Vulkan symbol naming in Ash.

File: ash-window/Cargo.toml [hunk 12]
```diff
@@ -18,13 +18,21 @@ rust-version = "1.64.0"
...
+
 [target.'cfg(any(target_os = "macos", target_os = "ios"))'.dependencies]
 raw-window-metal = "0.3"
 
 [dev-dependencies]
 winit = "0.28.0"
 ash = { path = "../ash", version = "0.37", default-features = false, features = ["linked"] }
 
+[features]
+default = [ "x11-dl" ]
```
Comment (MarijnS95) [inline ash-window/Cargo.toml:34]: Let's stick with the same formatting that the rest of this file uses:


```suggestion
default = ["x11-dl"]
```

File: ash-window/Cargo.toml [hunk 13]
```diff
@@ -18,13 +18,21 @@ rust-version = "1.64.0"
 ash = { path = "../ash", version = "0.37", default-features = false }
 raw-window-handle = "0.5"
 
+[target.'cfg(unix)'.dependencies]
+x11-dl = { version="2.21.0", optional=true }
+x11 = { version="2.21.0", features=["xlib"], optional=true }
```
Comment (MarijnS95) [inline ash-window/Cargo.toml:23]: Let's stick with the same formatting that the rest of this file uses:
```suggestion
x11-dl = { version = "2.21.0", optional = true }
x11 = { version = "2.21.0", features = ["xlib"], optional = true }
```

File: ash-window/src/lib.rs [hunk 14]
```diff
@@ -171,3 +171,102 @@ pub fn enumerate_required_extensions(
...
+        #[cfg(unix)]
+        RawDisplayHandle::Xcb(h) => unsafe {
+            // https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap34.html#platformQuerySupport_xcb
+            let ext = khr::XcbSurface::new(entry, instance);
+
+            let xcb = xcb::Connection::from_raw_conn(h.connection.cast());
+            let setup = xcb.get_setup();
+            let screen = setup.roots().nth(h.screen as usize).unwrap();
+            let visual = screen.root_visual();
+            xcb.into_raw_conn();
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:227]: For cleanliness, can we write `let connection = ...` here and pass that into `get_physical_device_xcb_presentation_support()` instead?

File: ash-window/src/lib.rs [hunk 15]
```diff
@@ -171,3 +171,102 @@ pub fn enumerate_required_extensions(
...
+        },
+        RawDisplayHandle::Windows(_) => unsafe {
+            // https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap34.html#platformQuerySupport_win32
+            let ext = khr::Win32Surface::new(entry, instance);
+            Ok(ext.get_physical_device_win32_presentation_support(
+                physical_device,
+                queue_family_index,
+            ))
+        },
+        #[cfg(unix)]
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:218]:
```suggestion
        #[cfg(feature = "xcb")]
```

File: ash-window/src/lib.rs [hunk 16]
```diff
@@ -171,3 +171,102 @@ pub fn enumerate_required_extensions(
...
+            let res = ext.get_physical_device_xcb_presentation_support(
+                physical_device,
+                queue_family_index,
+                h.connection,
+                visual,
+            );
+
+            Ok(res)
+        },
+        #[cfg(unix)]
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:238]:
```suggestion
        #[cfg(any(feature = "x11-dl", feature = "x11"))]
```

File: ash-window/src/lib.rs [hunk 17]
```diff
@@ -171,3 +171,102 @@ pub fn enumerate_required_extensions(
...
+        RawDisplayHandle::Xlib(h) => unsafe {
+            // https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap34.html#platformQuerySupport_xlib
+            let ext = khr::XlibSurface::new(entry, instance);
+
+            #[cfg(not(any(feature = "x11-dl", feature = "x11")))]
+            compile_error!("Exactly one of x11-dl or x11 must be enabled on unix");
+            #[cfg(all(feature = "x11-dl", feature = "x11"))]
+            compile_error!("Exactly one of x11-dl or x11 must be enabled on unix");
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:246]: Features should be additive, so if we can just pick one of the two crates instead of forcing them to be mutually exclusive?

File: ash-window/src/lib.rs (hunk 2 above)
Comment (MarijnS95) [reply ash-window/src/lib.rs]: > Bonus points for adding a cargo feature to use `x11` instead, as dynamic linking is a bit tidier than runtime loading on Linux, but it can be nice to avoid a buildtime dep on a foreign lib.

Quickly scrolling through the changes, it seems that this feature is now forcibly requiring either `x11` or `x11-dl`, or a compiler error will occur on _all of `unix`_. Same for `xcb` now being forced on all consumers of `ash-window`.

The `cfg` guards on those `RawDisplayHandle`s should be refactored to contain the optional libraries as features, otherwise `get_present_support()` won't provide support for x11/xcb but the other functions are not affected.

Then, the module documentation and function documentation should be extended to explain these features.

File: ash-window/src/lib.rs [hunk 18]
```diff
@@ -171,3 +171,102 @@ pub fn enumerate_required_extensions(
...
 }
+
+/// Query whether a `queue_family` of the given `physical_device` supports presenting to any
+/// surface that might be created. This function can be used to find a suitable
+/// [`vk::PhysicalDevice`] and queue family for rendering before a single surface is created.
+///
+/// This function can be a more useful alternative for [`vkGetPhysicalDeviceSurfaceSupportKHR`],
+/// which requires having an actual surface available before choosing a physical device.
+///
+/// For more information see [the vulkan spec on WSI integration][_querying_for_wsi_support].
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:182]: Once again:


```suggestion
/// For more information see [the Vulkan spec on WSI integration][_querying_for_wsi_support].
```

File: ash-window/src/lib.rs [hunk 19]
```diff
@@ -171,3 +171,102 @@ pub fn enumerate_required_extensions(
...
+            }
+
+            Ok(ext.get_physical_device_xlib_presentation_support(
+                physical_device,
+                queue_family_index,
+                h.display,
+                visual_id as _,
+            ))
+        },
+        // All other platforms mentioned in the vulkan spec don't
```
Comment (MarijnS95) [inline ash-window/src/lib.rs:268]: https://github.com/ash-rs/ash/pull/774#discussion_r1377297481

```suggestion
        // All other platforms mentioned in the Vulkan spec don't
```

**Comment by MarijnS95** [comment]
> I just noticed that `--no-default-features` seems to be ignored for running the winit example, I am not sure if I made a mistake or this is a cargo quirk.

How so? If I run this on your branch, which disables both `x11` and `x11-dl` (which I do not care about when using Wayland), the creates are disables as expected:

```
$ cargo r -p ash-window --example winit --no-default-features
...
error: Exactly one of x11-dl or x11 must be enabled on unix
   --> ash-window/src/lib.rs:244:13
    |
244 |             compile_error!("Exactly one of x11-dl or x11 must be enabled on unix");
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Though as said above, I don't think we should make this a compile error.

**Comment by Rob2309** [comment]
I now made every feature optional, but made x11-dl and xcb the default, so that every platform just works without enabling any features manually. A possible problem I see is that when x11/x11-dl is not enabled, the function will just return a generic EXTENSION_NOT_PRESENT error, which might be confusing. The same goes for xcb.

**Comment by Rob2309** [comment]
> > I just noticed that `--no-default-features` seems to be ignored for running the winit example, I am not sure if I made a mistake or this is a cargo quirk.
> 
> How so? If I run this on your branch, which disables both `x11` and `x11-dl` (which I do not care about when using Wayland), the creates are disables as expected:
> 
> ```
> $ cargo r -p ash-window --example winit --no-default-features
> ...
> error: Exactly one of x11-dl or x11 must be enabled on unix
>    --> ash-window/src/lib.rs:244:13
>     |
> 244 |             compile_error!("Exactly one of x11-dl or x11 must be enabled on unix");
>     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ```
> 
> Though as said above, I don't think we should make this a compile error.

I just found out that cargo seems to ignore `--no-default-features` when not explicitly specifying a package. So `cargo run --example winit --no-default-features` still uses default features, but `cargo run -p ash-window --example winit --no-default-features` does not... Weird cargo quirk I guess

**Comment by MarijnS95** [comment]
> I just found out that cargo seems to ignore `--no-default-features` when not explicitly specifying a package. So `cargo run --example winit --no-default-features` still uses default features, but `cargo run -p ash-window --example winit --no-default-features` does not... Weird cargo quirk I guess

I didn't expect `cargo run --example` to collect all examples from all members in the workspace automatically. TIL :)

**Comment by MarijnS95** [comment]
Cool stuff! Can we extend the example to showcase this API?

I hacked in this chunk to see the _type_ of window handle and print the result, but we should probably come up with something a bit nicer.

```rust
        dbg!(window.raw_display_handle());

        let &pdev = instance.enumerate_physical_devices()?.first().unwrap();

        let support = ash_window::get_present_support(
            &entry,
            &instance,
            pdev,
            0,
            window.raw_display_handle(),
        );
        dbg!(support);
```

And on that note, Rust naming says to not use `get_` prefixes; can we come up with a better name? `has_present_support(for_display_handle)()` perhaps?

**Comment by Rob2309** [comment]
> And on that note, Rust naming says to not use `get_` prefixes; can we come up with a better name? `has_present_support(for_display_handle)()` perhaps?

I would probably prefer keeping the get_prefix, as all the platform specific functions used also start with a get prefix. Ash itself also leaves the get_ prefix in place for many functions. But that is just my opinion

**Comment by MarijnS95** [comment]
> winit seems to not actually support xcb as far as I can tell.

I just ran into https://github.com/rust-windowing/winit/issues/3198

**Comment by Rob2309** [comment]
I added a simple loop to the example that prints info about every physical device.

**Comment by Rob2309** [comment]
> > winit seems to not actually support xcb as far as I can tell.
> 
> I just ran into [rust-windowing/winit#3198](https://github.com/rust-windowing/winit/issues/3198)

Since I can't get winit to use xcb, I might test the xcb code by creating a window with the xcb crate directly. Might take some time to get it right though...

**Comment by Rob2309** [comment]
I now added a basic xcb example, the xcb code seems to be working just fine. Not sure if the example should be kept though...

**Comment by MarijnS95** [comment]
> Since I can't get winit to use xcb, I might test the xcb code by creating a window with the xcb crate directly. Might take some time to get it right though...

No need to, it's supposed to a simple `winit` example. We can wait for them to implement `xcb`.

File: ash-window/examples/xcb.rs [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+            for i in 0..queue_families.len() {
+                let present_support = ash_window::get_present_support(
+                    &entry,
+                    &instance,
+                    dev,
+                    i as _,
+                    display_handle,
+                )?;
+                println!(
+                    "{dev_name}, queue {i} {} presenting to the surface",
```
Comment (MarijnS95) [inline ash-window/examples/xcb.rs:43]: Nit: "a" surface? There's no "the surface" to refer to here yet?

Comment (Rob2309) [reply ash-window/examples/xcb.rs:43]: Whoops it seems like I forgot that the entire purpose is to check before a surface exists :D

**Comment by Rob2309** [comment]
> > Since I can't get winit to use xcb, I might test the xcb code by creating a window with the xcb crate directly. Might take some time to get it right though...
> 
> No need to, it's supposed to a simple `winit` example. We can wait for them to implement `xcb`.

In that case, should I remove the example again? I used it to verify that the xcb code works, but I don't think it adds any value that the winit example does not.

**Comment by MarijnS95** [comment]
Hmm you have already written it, but it also adds some extra maintenance burden to keep it up-to-date and functional.

File: ash-window/Cargo.toml [hunk 21]
```diff
@@ -18,13 +18,26 @@ rust-version = "1.64.0"
...
 [target.'cfg(any(target_os = "macos", target_os = "ios"))'.dependencies]
 raw-window-metal = "0.3"
 
 [dev-dependencies]
 winit = "0.28.0"
 ash = { path = "../ash", version = "0.37", default-features = false, features = ["linked"] }
+xcb = { version = "1.2.2" }
+
+[features]
+default = ["x11-dl", "xcb"]
```
Comment (Ralith) [inline ash-window/Cargo.toml:35]: Requiring `xcb` by default means we need libxcb present at buildtime, right? While probably most environments will have this, disabling it will make it easier for downstream users to get CI working in minimal VMs and such.

**Review by Ralith — APPROVED** [review]
Implementation LGTM overall, thanks!

**Review by MarijnS95 — APPROVED** [review]
Looks like I forgot to submit this outstanding review. Don't forget to address https://github.com/ash-rs/ash/pull/774/files#r1378560481 :)

File: ash-window/Cargo.toml (hunk 21 above)
Comment (Rob2309) [reply ash-window/Cargo.toml:35]: The reason for defaulting this is to have the least surprising behavior by default. I dont think the default should be to not support a very broadly used platform.

File: ash-window/Cargo.toml [hunk 22]
```diff
@@ -18,13 +18,26 @@ rust-version = "1.64.0"
...
 
 [[example]]
 name = "winit"
 required-features = ["ash/linked"]
+
+[[example]]
+name = "xcb"
+required-features = ["ash/linked"]
```
[the remark this answers is not in the document]
Comment (Rob2309) [reply ash-window/Cargo.toml:43]: I am not quite sure whether normal dependencies are available in examples. Will test this and change if possible.

**Comment by Rob2309** [comment]
> Looks like I forgot to submit this outstanding review. Don't forget to address https://github.com/ash-rs/ash/pull/774/files#r1378560481 :)

I can add this to the changelog, I was just hesitant because of the separate PR point.

File: ash-window/Cargo.toml (hunk 21 above)
Comment (Ralith) [reply ash-window/Cargo.toml:35]: We know winit doesn't use xcb. Where in the raw-window-handle ecosystem is xcb broadly used?

**Comment by Rob2309** [comment]
The forced dev-dependency is now removed, it seems to just work with the regular optional dependency.
I also added entries to both the ash-window and ash changelog reflecting the changes here. Is there any special marking that the breaking changes should have?

File: ash-window/Cargo.toml (hunk 21 above)
Comment (Rob2309) [reply ash-window/Cargo.toml:34]: Good point, however if winit at some point decides to switch to xcb, it should be defaulted again...

Comment (Ralith) [reply ash-window/Cargo.toml:34]: I think winit will want to avoid adding foreign build-time dependencies for the same reason.

**Comment by Rob2309** [comment]
Is there anything left to do here? I must admit I am not quite sure about the status of this PR anymore...

**Comment by Ralith** [comment]
Triage: This is still desirable, and is nearly ready. I've gone through and resolved threads which have been addressed.

Remaining concerns:
- enabling `xcb` by default on Unix platforms may be a poor experience, especially on e.g. macos. Also possibly not desired in Wayland-only environments.
- Needs a rebase.
